File: leetcode/C++/utf-8-validation.cpp
class Solution {
public:
    vector<string> v;
    void trans(int &d){
        string s;
        for(int i=128; i>0; i>>=1){
            if((d&i)) s.push_back('1');
            else s.push_back('0');
        }
        v.push_back(s);
    }
    bool validUtf8(vector<int>& data) {
        int sz = data.size(), state=0, bit=0;
        for(int &t : data) trans(t);
        for(int i=0; i<sz; ++i){
            for(int j=0; j<8; ++j){
                if(state==0){
                    if(v[i][0]=='0') break;
                    else if(v[i][j]=='1') ++bit;
                    else {
                        if(bit<=1 || bit>4) return false;
                        ++state; --bit;
                        break;
                    }
                }
                else {
                    if(v[i][0]=='1' && v[i][1]=='0') --bit;
                    else return false;
                    if(bit==0) state=0;
                    break;
                }
            }
        }
        return bit==0;
    }
};

File: leetcode/Python/utf-8-validation.py
class Solution:
    def validUtf8(self, data: List[int]) -> bool:
        v, state, bit = [], 0, 0
        def _trans(n):
            s, i= [], 128;
            while i>0:
                if i&n > 0:
                    s.append(1)
                else:
                    s.append(0)
                i>>=1
            v.append(s)
        for d in data:
            _trans(d)
        for d in v:
            for i in range(8):
                if state==0:
                    if d[0]==0:
                        break
                    elif d[i]==1:
                        bit+=1
                    else:
                        if bit<=1 or bit>4:
                            return False
                        state=1
                        bit-=1
                        break
                else:
                    if d[0]==1 and d[1]==0:
                        bit-=1
                    else:
                        return False
                    if bit==0:
                        state=0
                    break
        print(v)
        return bit==0